### PR TITLE
PluginComponent contributes description to -Xshow-phases. (Fixes SI-6446)

### PR DESCRIPTION
**Pull request by som-snytt**
In Global, SubComponent is called a phase descriptor, but it doesn't actually have a description. (Phase itself does.) This fix adds a description to PluginComponent so that plugins can describe what they do in -Xshow-phases.

To avoid chaos, as per the formatting comments in Global.scala, -Xshow-phases output is restricted to 80 columns.  As an aid, inlineExceptionHandlers is renamed to inlineHandlers.

Phase descriptions are provided for the CPS components.

A minimal test is provided for -Xshow-phases with a custom plugin.  Plugin loading and partest are modified so that a test can provide a plugin.  (Plugin loading accepts an exploded jar, and partest replaces "dot" or cwd with output dir.  This means a test can contribute just one plugin.)

(One could imagine testing a compiler component by supplying a test phase that creates possibly screwy trees, for example to test robustness.)

Moves test run/t4351.scala from the pos test set.

Footnote: Continuations tests could be mainstreamed, by adding .flags.

review by: @paulp who cares

### CODE REVIEW

File: test/files/run/programmatic-main.check [hunk 1]
```diff
@@ -1,31 +1,31 @@
...
+   posterasure  15  clean up erased inline classes
+      lazyvals  16  allocate bitmaps, translate lazy vals into lazified defs
+    lambdalift  17  move nested functions to top level
+  constructors  18  move field definitions into constructors
+       flatten  19  eliminate inner classes
+         mixin  20  mixin composition
+       cleanup  21  platform-specific cleanups, generate reflective calls
+         icode  22  generate portable intermediate code
+       inliner  23  optimization: do inlining
+inlineHandlers  24  optimization: inline exception handlers
```
Comment (paulp) [inline test/files/run/programmatic-main.check]: That "H" is now the only big boy letter among all the phase names.

File: src/compiler/scala/tools/nsc/plugins/PluginDescription.scala [hunk 2]
```diff
@@ -45,31 +40,62 @@ abstract class PluginDescription {
...
-      case Some(str) => str
+    // check the top-level tag
+    xml match {
+      case <plugin>{_*}</plugin>  => extracted
+      case _                      => None
     }
+  }
+
+/*
+  import scala.reflect.runtime.{currentMirror => cm, universe => uni}
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/plugins/PluginDescription.scala]: In the interests of keeping the number of aliases to a minimum, I believe the standard for universe is "ru".

File: test/files/run/programmatic-main.check (hunk 1 above)
Comment (som-snytt) [inline test/files/run/programmatic-main.check]: Wow, you really do care!  Good catch.  Luckily, "h" is also tall, so it should be legible despite length.

File: src/compiler/scala/tools/nsc/plugins/PluginDescription.scala (hunk 2 above)
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/plugins/PluginDescription.scala]: Right. That is in a comment that answers, is this easier to do reflectively, and the answer is not really.  (You might remember someone asked on the ML can I avoid using string names when constructing objects from amorphous data.)  Shall I just delete the comment?

The fun part of that trial was supplying ctor defaults like PluginMain for the class, but that was just a lark.  No one is clamoring to reduce boilerplate for plugins.

File: test/files/run/programmatic-main.check (hunk 1 above)
Comment (som-snytt) [inline test/files/run/programmatic-main.check]: I was sanguine at first, but I'd forgotten that the string is scattered about.  My original edit was to override phaseName in global; but maybe with lazy val phase objects, there'd be a problem at assembly; second idea was just an enum of phase names and descriptions, so at least they're all in one place.  But I haven't read all the comments in firstPhase yet.

I didn't even realize inlineHandlers is a setting.  I can't be trusted.

**Comment by som-snytt** [comment]
There are additional cogent rejoinders buried behind the outdated diffs.  I forgot that part about pushing.

**Comment by som-snytt** [comment]
This is the partestable version of 
https://github.com/scala/scala/issues/1419

If it's not deemed to pull its weight, I'll kill it before it's no longer mergeable.  I already see issues with the issue.

@gkossakowski or @paulp might ask for a more substantial contribution to consume review-by resources.

**Comment by som-snytt** [comment]
Reopening with rebased commit.  Writing a plugin is so much more fun with a custom description.  I understand that it may be a slow process, but worth the wait.  Also, quite a lot of elven labor frees up in about a week, so maybe it will get some cheer.  (I wasn't actually referring to the elves who appear in the workshop every night to work on 2.10-final, but maybe they could help, too.)

I added more tests and tweaked some code. It's too bad the formatting logic is in global and not in reporter.

**Comment by paulp** [comment]
This looks great to me.

**Comment by paulp** [comment]
"It's too bad the formatting logic is in global and not in reporter."

We have the power to move stuff.

**Comment by som-snytt** [comment]
I remember looking at moving it, and I've forgotten the detail, but it had to do with how the "info" return from compiler interacts.  (Or I don't remember.) So if desired, I would refactor that separately.

**Comment by som-snytt** [comment]
Thanks.  I noticed the sbt stack trace with the plugin classloading issue didn't include this commit (whew?).  My New Year's resolution is to always have a PR in the review queue.  Excuse the split infinitive.  Happy Holidays.

**Pull request merged**